src/field_theory/splitting_field.lean

```diff
@@ -94,61 +95,405 @@ by rw [nat_degree_remove_factor, hfn, n.add_sub_cancel]
-ih f.remove_factor
+def splitting_field_aux : Π (n : ℕ) {K : Type u} [field K], by exactI Π (f : K[X]), Type u
+| 0 K fK f := K
+| (n + 1) K fK f := by exactI splitting_field_aux n f.remove_factor
 
 namespace splitting_field_aux
 
 theorem succ (n : ℕ) (f : K[X]) :
   splitting_field_aux (n+1) f = splitting_field_aux n f.remove_factor := rfl
 
-instance field (n : ℕ) : Π {K : Type u} [field K], by exactI
-  Π {f : K[X]}, field (splitting_field_aux n f) :=
-nat.rec_on n (λ K _ _, ‹field K›) $ λ n ih K _ f, ih
+section lift_instances
 
-instance inhabited {n : ℕ} {f : K[X]} :
-  inhabited (splitting_field_aux n f) := ⟨37⟩
+/-! ### Instances on `splitting_field_aux`
 
-/-
-Note that the recursive nature of this definition and `splitting_field_aux.field` creates
-non-definitionally-equal diamonds in the `ℕ`- and `ℤ`- actions.
-```lean
-example (n : ℕ) {K : Type u} [field K] {f : K[X]} (hfn : f.nat_degree = n) :
-    (add_comm_monoid.nat_module : module ℕ (splitting_field_aux n f hfn)) =
-  @algebra.to_module _ _ _ _ (splitting_field_aux.algebra n _ hfn) :=
-rfl  -- fails
-```
-It's not immediately clear whether this _can_ be fixed; the failure is much the same as the reason
-that the following fails:
-```lean
-def cases_twice {α} (a₀ aₙ : α) : ℕ → α × α
-| 0 := (a₀, a₀)
-| (n + 1) := (aₙ, aₙ)
-
-example (x : ℕ) {α} (a₀ aₙ : α) : (cases_twice a₀ aₙ x).1 = (cases_twice a₀ aₙ x).2 := rfl  -- fails
-```
-We don't really care at this point because this is an implementation detail (which is why this is
-not a docstring), but we do in `splitting_field.algebra'` below. -/
-instance algebra (n : ℕ) : Π (R : Type*) {K : Type u} [comm_semiring R] [field K],
-  by exactI Π [algebra R K] {f : K[X]},
-    algebra R (splitting_field_aux n f) :=
-nat.rec_on n (λ R K _ _ _ _, by exactI ‹algebra R K›) $
-         λ n ih R K _ _ _ f, by exactI ih R
+In order to avoid diamond issues, we have to be careful to define each field of
+an instance of `splitting_field_aux` by recursion, rather than defining the
+whole structure by recursion at once.
+
+We do this in three steps: first define an `add_comm_group` instance to put a `distrib_smul`
+on (which is what is needed to push scalar multiplication through the recursive step),
+then a `comm_ring` instance, and finally a `field` instance.
+-/
+
+/-- Splitting fields have a zero. -/
+protected def zero (n : ℕ) : Π {K : Type u} [field K], by exactI Π {f : K[X]},
+  splitting_field_aux n f :=
+nat.rec_on n (λ K fK f, by exactI @has_zero.zero K _) (λ n ih K fK f, ih)
+
+/-- Splitting fields have an addition. -/
+protected def add (n : ℕ) : Π {K : Type u} [field K], by exactI Π {f : K[X]},
+  splitting_field_aux n f → splitting_field_aux n f → splitting_field_aux n f :=
+nat.rec_on n (λ K fK f, by exactI @has_add.add K _) (λ n ih K fK f, ih)
+
+/-- `K` adjoined `n` roots of `f : K[X]` has addition and zero.
+
+We need this instance to define the `distrib_smul` instance below; a `distrib_smul`
```

Comment:
Maybe something like
```suggestion
We stop via this instance on our path to the `field` instance in order to define the `distrib_smul` instance below; a `distrib_smul`
```
We need the instance anyway, the interesting thing is that we can't go all the way to `field` straight away